### PR TITLE
Del

### PR DESCRIPTION
**Pull request by Yggdrasillll**
add focus controller

### CODE REVIEW

File: dbux-graph-client/src/components/ContextNode.js [hunk 1]
```diff
@@ -1,10 +1,14 @@
-import { createPopper } from '@popperjs/core';
+import popperManeger from '../popperManager';
```
Comment (Domiii) [inline dbux-graph-client/src/components/ContextNode.js:1]: typo: `popperManeger`

File: dbux-graph-client/src/components/GraphRoot.js [hunk 2]
```diff
@@ -1,31 +1,46 @@
 import createPanzoom from 'panzoom';
 import ClientComponentEndpoint from '../componentLib/ClientComponentEndpoint';
 import { compileHtmlElement } from '@/util/domUtil';
+import popperManeger from '../popperManager';
```
Comment (Domiii) [inline dbux-graph-client/src/components/GraphRoot.js:4]: typo: `popperManeger`

File: dbux-graph-client/src/components/GraphRoot.js [hunk 3]
```diff
@@ -74,6 +90,7 @@ class GraphRoot extends ClientComponentEndpoint {
 
     panzoom.on('zoom', (e) => {
       // this._repaint();
+      popperManeger.update();
```
Comment (Domiii) [inline dbux-graph-client/src/components/GraphRoot.js:93]: typo

File: dbux-graph-client/src/components/GraphRoot.js [hunk 4]
```diff
@@ -1,31 +1,46 @@
...
+        </div>   
       </div>
     `);
   }
 
+  setupEl() {
+    this.panzoom = this.initPanZoom(this.els.body);
+    
+    // hackfix: make popperEl global for now
+    window._popperEl = this.els.toolTip;
```
Comment (Domiii) [inline dbux-graph-client/src/components/GraphRoot.js:39]: create a new `controller` instead

File: dbux-graph-client/src/components/ContextNode.js [hunk 5]
```diff
@@ -33,60 +37,20 @@ class ContextNode extends ClientComponentEndpoint {
   update() {
     const {
       displayName,
+      applicationId,
       context: { contextId, staticContextId }
     } = this.state;
 
 
-    this.el.id = `context_${contextId}`;
+    this.el.id = `application_${applicationId}-context_${contextId}`;
```
Comment (Domiii) [inline dbux-graph-client/src/components/ContextNode.js:45]: don't use `id` here, replace with proper `controller` instead:
* on `Host`, use `contextNode = this.context.graphRoot.getContextNode(context)`
* then use that `contextNode` to (1) slide, (2) highlight, (3) reveal (all three features 功能) can adopt this strategy

File: dbux-graph-client/src/popperManager.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import { createPopper } from '@popperjs/core';
+
+class PopperManeger {
```
Comment (Domiii) [inline dbux-graph-client/src/popperManager.js:3]: typo 👯

File: dbux-graph-client/src/popperManager.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    this.create(button, tooltip);
+  }
+
+  hide = (tooltip) => {
+    tooltip.removeAttribute('data-show');
+    this.destroy();
+  }
+
+  update = () => {
+    if (this.popperInstance) {
```
Comment (Domiii) [inline dbux-graph-client/src/popperManager.js:40]: can use "conditional chaining" instead: `this.popperInstance?.update()`

File: dbux-graph-host/src/components/controllers/Highlighter.js [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import HostComponentEndpoint from 'dbux-graph-host/src/componentLib/HostComponentEndpoint';
+import { getClosestPackageJsonNameOrPath } from '../../../../dbux-data/src/util/nodeUtil';
```
Comment (Domiii) [inline dbux-graph-host/src/components/controllers/Highlighter.js:2]: ehm... that's probably a mistake?

File: dbux-graph-client/src/components/GraphRoot.js (hunk 4 above)
Comment (MiccWan) [reply dbux-graph-client/src/components/GraphRoot.js:39]: i can help with this

File: dbux-graph-host/src/components/controllers/Highlighter.js (hunk 8 above)
Comment (MiccWan) [reply dbux-graph-host/src/components/controllers/Highlighter.js:2]: Checked unused and removed.

**Pull request merged**